### ISSUE TITLE
Serialization/Deserialization

### ISSUE THREAD

**Issue by cakey** [Stoeoef/spade#15]
Hey,
I'd be interested in serialization / deserialization to disk - any suggestions on where I should start to add this functionality?

Thanks,
Ben

**Comment by Stoeoef** [comment]
Hey Ben,

thanks for your interest! Sorry for my late reply, I only got back to spade recently.
serialization / deserialization is a great idea and shouldn't be too difficult. So, if you are still interested, here are some starting points:
- The main data structures are the r-tree and the delaunay triangulation, adding (de)serialization to these should be the focus.
- The types in the primitive module should also support serialization.
- I never had to implement serialization in rust, but implementing the appropriate serde traits might be the most useful serialization solution. Please correct me if I'm wrong.
- As far as I've read through serde, a simple `#[derive(Serialize, Deserialize)]` at many places might do the trick. The compiler would guide you through that process.
- As an advanced feature: You could implement a general serialization of the `PointN` trait. This would...
  - allow spade to serialize points that do not implement `Serialize`
  - even convert between trees and triangulations of different point types.

**Comment by adeschamps** [comment]
To avoid breaking code that uses types which don't implement `Serialize` and `Deserialize`, perhaps this should be done with a feature flag. In a lot of cases, that would just mean writing this:

```rust
#[cfg(feature = "serde")]
#[derive(Serialize, Deserialize)]
struct Foo { /* ... */ }
```

**Comment by Stoeoef** [comment]
Implemented in version v1.4 . Enabled with the `serde_serialize` feature
